### PR TITLE
cmake parity updates

### PR DESCRIPTION
**Pull request by compnerd**
This updates the CMake build to bring it to parity with the autotools build.  It also cleans up some of the custom list management for more modern approaches.

### CODE REVIEW

**Comment by compnerd** [comment]
CC @das @MadCoder @llvm-beanz @gottesmm

**Comment by das** [comment]
very cool, thanks!

BTW, I've made some progress on getting the CMake build to work on Darwin (against submodules of the dependent projects under github.com/apple/darwin*), I'll have to rebase on top of this and see if that solves some of the problems I was running into

**Comment by compnerd** [comment]
@das awesome!  Really looking forward to seeing this become the one build system for libdispatch.  I am able to build with this change on linux and all tests pass.  If you have specific things that I can try to address in the build system, to help in getting everything switched over, let me know!

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -90,9 +52,57 @@ add_library(dispatch
...
+                      swift/Dispatch.swift
+                      swift/IO.swift
+                      swift/Private.swift
+                      swift/Queue.swift
+                      swift/Source.swift
+                      swift/Time.swift
+                      swift/Wrapper.swift
+                    CFLAGS
+                      -fblocks
+                      -fmodule-map-file=${CMAKE_SOURCE_DIR}/dispatch/module.modulemap
```
Comment (das) [inline src/CMakeLists.txt]: this is a pre-existing bug but this is the linux modulemap, if we ever want this part to work on Darwin this path will need to be conditionalized per platform

**Pull request merged**